lib/Analysis/ControlDependenceAnalysis.cpp

```diff
@@ -0,0 +1,298 @@
+#include "mlir/Analysis/CFGLoopInfo.h"
+#include "mlir/IR/Dominance.h"
+#include "llvm/Support/Debug.h"
+
+using namespace mlir;
+using namespace mlir::func;
+using namespace dynamatic;
+
+#define DEBUG_TYPE "cdg-analysis"
+
+using PathInDomTree = SmallVector<DominanceInfoNode *, 4>;
+
+void ControlDependenceAnalysis::identifyAllControlDeps(FuncOp &funcOp) {
+  Region &funcReg = funcOp.getRegion();
+  DenseMap<Block *, llvm::SmallVector<Block *, 4>> control_deps_map;
+
+  // Initialize the all_control_deps_maps by creating an entry for every block
+  // constituting the region
+  for (Block &block : funcReg.getBlocks()) {
+    SmallVector<Block *, 4> deps;
+    control_deps_map.insert(std::make_pair(&block, deps));
+  }
+
+  // functions made up of 1 basic block do not have any dependencies
+  if (funcReg.hasOneBlock())
+    return;
+
+  // Get information about post-dominance
+  PostDominanceInfo postDomInfo;
+  // Get the post-dominance tree
+  llvm::DominatorTreeBase<Block, true> &postDomTree =
+      postDomInfo.getDomTree(&funcReg);
+
+  // Loop over the control flow edges connnecting the different blocks of this
+  // region
+  for (Block &block : funcReg.getBlocks()) {
+    for (Block *block_succ : block.getSuccessors()) {
+      if (!postDomInfo.properlyPostDominates(block_succ, &block)) {
+        Block *least_common_anc =
+            postDomInfo.findNearestCommonDominator(block_succ, &block);
+
+        if (least_common_anc == &block)
+          // Loop case
+          control_deps_map[&block].push_back(&block);
+
+        // All nodes in the post-dominator tree on the path from the
+        // "least_common_anc" to "block_succ" (including "block_succ") should
+        // be control dependent on "block"
+        // Easy case of block_succ
+        control_deps_map[block_succ].push_back(&block);
+        // Enumerate all paths between "least_common_anc" and "block_succ"
+        SmallVector<PathInDomTree, 4> allPathsFromLeastCommonAncToBlockSucc;
+        enumeratePathsInPostDomTree(least_common_anc, block_succ, &funcReg,
+                                    &postDomTree,
+                                    &allPathsFromLeastCommonAncToBlockSucc);
+
+        for (PathInDomTree path : allPathsFromLeastCommonAncToBlockSucc)
+          for (DominanceInfoNode *domInfo : path) {
+            Block *blockInPath = domInfo->getBlock();
+
+            // Skip the nodes that we have already taken care of above
+            if (blockInPath == least_common_anc || blockInPath == &block ||
+                blockInPath == block_succ)
+              continue;
+
+            // Add block to the control dependencies only if it is not already
+            // there
+            if (std::find(control_deps_map[blockInPath].begin(),
+                          control_deps_map[blockInPath].end(),
+                          &block) == control_deps_map[blockInPath].end())
+              control_deps_map[blockInPath].push_back(&block);
```

Comment:
If that's the logic you want, then it seems that `control_deps_map` should be a `SmallVector<DenseMap<Block*, DenseSet<Block*>>>` instead, which removes the need for this linear-time check here before inserting the block. If iteration order is important in the set of blocks, then use a `SetVector<Block*>` instead.

In any case, a set carries the meaning that all elements are unique, whereas a vector does not.